upgrade-tests/src/test/java/io/zeebe/test/UpgradeTestCase.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.0. You may not use this file
+ * except in compliance with the Zeebe Community License 1.0.
+ */
+package io.zeebe.test;
+
+import io.zeebe.client.ZeebeClient;
+import io.zeebe.model.bpmn.BpmnModelInstance;
+import io.zeebe.util.collection.Tuple;
+import java.util.Map;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+final class UpgradeTestCase {
+  private TestCaseBuilder builder;
+
+  private UpgradeTestCase(final TestCaseBuilder builder) {
+    this.builder = builder;
+  }
+
+  static TestCaseBuilder builder() {
+    return new TestCaseBuilder();
+  }
+
+  long setUp(final ZeebeClient client) {
+    builder.deployWorkflow.accept(client);
+    return builder.createInstance.apply(client);
+  }
+
+  Long runBefore(final ContainerStateRule state) {
+    return builder.before.apply(state);
+  }
+
+  void runAfter(final ContainerStateRule state, final long wfInstanceKey, final long key) {
+    builder.after.accept(state, wfInstanceKey, key);
+  }
+
+  static class TestCaseBuilder {
+    private Consumer<ZeebeClient> deployWorkflow = c -> {};
+    private Function<ZeebeClient, Long> createInstance = c -> -1L;
+    private Function<ContainerStateRule, Long> before = r -> -1L;
+    private TriConsumer<ContainerStateRule, Long, Long> after = (r, wfKey, k) -> {};
+
+    TestCaseBuilder deployWorkflow(final BpmnModelInstance model) {
+      return deployWorkflow(new Tuple<>(model, UpgradeTest.PROCESS_ID));
+    }
+
+    @SafeVarargs
+    final TestCaseBuilder deployWorkflow(final Tuple<BpmnModelInstance, String>... models) {
+      for (final Tuple<BpmnModelInstance, String> model : models) {
+        deployWorkflow =
+            deployWorkflow.andThen(
+                client ->
+                    client
+                        .newDeployCommand()
+                        .addWorkflowModel(model.getLeft(), model.getRight() + ".bpmn")
+                        .send()
+                        .join());
+      }
+      return this;
+    }
+
+    TestCaseBuilder createInstance() {
+      createInstance =
+          client ->
+              client
+                  .newCreateInstanceCommand()
+                  .bpmnProcessId(UpgradeTest.PROCESS_ID)
+                  .latestVersion()
+                  .variables(Map.of("key", "123"))
```

Comment:
Without any comment, it is not clear that these variables are required by the message correlation tests. We could pass the variables from the test (where they are needed) or add at least a comment.